modules/alloy/src/main/ts/ephox/alloy/dom/Reconcile.ts

```diff
@@ -0,0 +1,76 @@
+import { Arr, Obj } from '@ephox/katamari';
+import { Attribute, Classes, Css, Html, SugarElement, Value } from '@ephox/sugar';
+
+import { DomDefinitionDetail } from './DomDefinition';
+import { patchDomChildren } from './Patching';
+
+interface KeyValueDiff {
+  readonly toSet: Record<string, string>;
+  readonly toRemove: string[];
+}
+
+const diffKeyValueSet = (newObj: Record<string, string>, oldObj: Record<string, string>): KeyValueDiff => {
+  const newKeys = Obj.keys(newObj);
+  const oldKeys = Obj.keys(oldObj);
+  const toRemove = Arr.difference(oldKeys, newKeys);
+  const toSet = Obj.bifilter(newObj, (v, k) => {
+    return !Obj.has(oldObj, k) || v !== oldObj[k];
+  }).t;
+
+  return { toRemove, toSet };
+};
+
+const reconcileToDom = (definition: DomDefinitionDetail, obsoleted: SugarElement<Element>): SugarElement<Element> => {
+  const { class: clazz, style, ...existingAttributes } = Attribute.clone(obsoleted);
+  const { toSet: attrsToSet, toRemove: attrsToRemove } = diffKeyValueSet(definition.attributes, existingAttributes);
+
+  const updateAttrs = () => {
+    Arr.each(attrsToRemove, (a) => Attribute.remove(obsoleted, a));
+    Attribute.setAll(obsoleted, attrsToSet);
+  };
+
+  const existingStyles = Css.getAllRaw(obsoleted);
+  const { toSet: stylesToSet, toRemove: stylesToRemove } = diffKeyValueSet(definition.styles, existingStyles);
+  const updateStyles = () => {
+    Arr.each(stylesToRemove, (s) => Css.remove(obsoleted, s));
+    Css.setAll(obsoleted, stylesToSet);
+  };
+
+  const existingClasses = Classes.get(obsoleted);
+  const classesToRemove = Arr.difference(existingClasses, definition.classes);
+  const classesToAdd = Arr.difference(definition.classes, existingClasses);
+
+  const updateHtml = (html: string) => {
+    Html.set(obsoleted, html);
```

Comment:
Yeah, the original spike has some really strange code in it. I see you've removed that which is good. Hopefully, it wasn't trying to cover some weird edge case. 

Your test coverage is probably pretty good, so it should be fine.

Reply:
Yeah, all that logic was for handling the case where it had mixed children from what I could tell. That just really can't be patched (or at least not without jumping through a heap of hoops) so I've just instead made mixed children specs unable to be patched (we rarely ever do it and I'd prefer to get rid of being able to do that anyways).